src/darwin/Framework/CHIP/MTRDevice.mm

```diff
@@ -995,47 +999,102 @@ - (NSArray *)_getAttributesToReportWithReportedValues:(NSArray<NSDictionary<NSSt
     return attributesToReport;
 }
 
+// If value is non-nil, associate with expectedValueID
+// If value is nil, remove only if expectedValueID matches
+- (void)_setExpectedValue:(NSDictionary<NSString *, id> *)expectedAttributeValue
+             attributePath:(MTRAttributePath *)attributePath
+            expirationTime:(NSDate *)expirationTime
+         shouldReportValue:(BOOL *)shouldReportValue
+    attributeValueToReport:(NSDictionary<NSString *, id> **)attributeValueToReport
+           expectedValueID:(uint64_t)expectedValueID
+{
+    os_unfair_lock_assert_owner(&self->_lock);
+
+    *shouldReportValue = NO;
+
+    NSArray * previousExpectedValue = _expectedValueCache[attributePath];
+    if (previousExpectedValue) {
+        if (expectedAttributeValue
+            && ![self _attributeDataValue:expectedAttributeValue
+                       isEqualToDataValue:previousExpectedValue[MTRDeviceExpectedValueFieldValueIndex]]) {
+            // Case where new expected value overrides previous expected value - report new expected value
+            *shouldReportValue = YES;
+            *attributeValueToReport = expectedAttributeValue;
+        } else if (!expectedAttributeValue
+            && ![self _attributeDataValue:previousExpectedValue[MTRDeviceExpectedValueFieldValueIndex]
+                       isEqualToDataValue:_readCache[attributePath]]) {
+            // Case of removing expected value that is different than read cache - report read cache value
+            *shouldReportValue = YES;
+            *attributeValueToReport = _readCache[attributePath];
+        }
+    } else {
+        if (expectedAttributeValue
+            && ![self _attributeDataValue:expectedAttributeValue isEqualToDataValue:_readCache[attributePath]]) {
+            // Case where new expected value is different than read cache - report new expected value
+            *shouldReportValue = YES;
+            *attributeValueToReport = expectedAttributeValue;
+        }
+
+        // No need to report if new and previous expected value are both nil
+    }
+
+    if (expectedAttributeValue) {
+        _expectedValueCache[attributePath] = @[ expirationTime, expectedAttributeValue, @(expectedValueID) ];
+    } else if (previousExpectedValue) {
+        // Remove previous expected value only if it's from the same setExpectedValues operation
+        NSNumber * previousExpectedValueID = previousExpectedValue[MTRDeviceExpectedValueFieldIDIndex];
+        if (previousExpectedValueID.unsignedLongLongValue == expectedValueID) {
```

Comment:
If this tests false, shouldn't we reset `*shouldReportValue` to `NO`?   Or check for this condition earlier or something....

As in, perhaps we should check for this condition pretty early on and take an early return if it holds, so that we don't have to worry about it here or in the "figure out value to report" logic above.

Reply:
Hmm you're right. I'll move this logic.

Reply:
Fixed in latest commit.